WikiWatershed/model-my-watershed
# HydroShare Copying: Generate project Snapshot and export to HydroShare

## Overview

Adds a method to the project model to generate a JSON snapshot of the entire model state, including a `snapshot_at` key with a timestamp of when the snapshot was made. This is included in the HydroShare Export as `mmw_project_snapshot.json`, to be read in at a later stage.

This snapshot is a complete representation of the project in MMW.

Connects #2999 

### Demo

https://beta.hydroshare.org/resource/e98d11dfc9a04b5689476665545fd3a6/

![image](https://user-images.githubusercontent.com/1430060/47166250-b747de00-d2c9-11e8-9b13-c0bbe84304ac.png)

```bash
$ jq '.id, .name, .snapshot_at' < mmw_project_snapshot.json

37
"HydroShare Copying"
"2018-10-18T15:22:43.107Z"
```

### Notes

I considered excluding the Area of Interest from the snapshot since it is already exported in its own file, and we'll most likely only be reading the project id from the snapshot file for now. But this doesn't seem to add a significant delay to the export, and maintains a complete representation of an MMW project, so I've kept it in for now.

## Testing Instructions

* Check out this branch and `bundle`
* Find the HydroShare Beta Client Secret Key in LastPass and populate `/etc/mmw.d/env/MMW_HYDROSHARE_SECRET_KEY` in the `app` VM and restart `mmw-app` service
* Go to [:8000/](http://localhost:8000/) and log in. Go to your account and under linked accounts, unlink from HydroShare if you are already linked. Then re-link to HydroShare. (You'll need a HydroShare Beta account to do this, you can make one if you don't have it already).
* Make a project and export it to HydroShare. Ensure the HydroShare export contains the `mmw_project_snapshot.json` file and it is an accurate representation of the project (has the right ID and name).
